### PR TITLE
[Breaking change] Apivore v1

### PR DESCRIPTION
**Pull request by gingermusketeer**

### CODE REVIEW

**Comment by gingermusketeer** [comment]
This is a work in progress. Need to be cleaned up a bit. Feedback welcome

File: lib/apivore/swagger_checker.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      mappings[path].has_key?(method)
+    end
+
+    def has_response_code_for_path?(path, method, code)
+      mappings[path][method].has_key?(code.to_s)
+    end
+
+    def has_matching_document_for(path, method, code, body)
+      JSON::Validator.fully_validate(
+        swagger, body, fragment: fragment(path, method, code), strict: true
```
Comment (hornc) [inline lib/apivore/swagger_checker.rb]: strict should not be true here

Comment (gingermusketeer) [inline lib/apivore/swagger_checker.rb]: I wondered about that. Thanks

**Comment by hornc** [comment]
Hi @gingermusketeer , I am concerned that these changes look like they are just testing the contents of the swagger documentation, when the original intent of this tool was to generate and run tests against the actual api using the swagger documentation.

**Comment by hornc** [comment]
To clarify my comment above, I'm basing my concern on the updated request_spec.rb test that now does not appear to make any requests to the API, where previously it made one request for each path, it just makes assertions against the swagger doc.

It seems to me that there we want a better way to test our swagger documentation? Apivore originally worked with the assumption that the swagger documentation was generally correct, and only provided a single basic schema validation pass, then used that to test that the API responded as documented.

If there was a mis-match, apivore would try to state what the mismatch was, but the concept of 'line numbers' isn't really applicable in many cases if a whole route or response does not exist in the api. It is intended to run as a request spec.

Maybe there is a misunderstanding of what apivore is trying to do, and a gap in what else we need?

**Comment by gingermusketeer** [comment]
@hornc It still interacts with the actual api. See https://github.com/westfieldlabs/apivore/pull/40/files#diff-62a36f1ebf6812e74bd31f41f5b85b85R21

**Comment by hornc** [comment]
Ah, thanks @gingermusketeer, I see that now. I was thrown by the expectation verb `document`
I accept that it is hard to think of a word that covers it all. I originally went with `validates` as is was sufficiently vague to cover everything with "validates that the swagger correctly documents the api"

**Comment by gingermusketeer** [comment]
@hornc This PR is a proof of concept. I was not sure if it was possible to do this before i started. Happy to change any wording

File: lib/apivore/document.rb [hunk 2]
```diff
@@ -0,0 +1,95 @@
+require 'action_controller'
+require 'action_dispatch'
+
+module Apivore
+  class Document
```
Comment (gwagener) [inline lib/apivore/document.rb]: I think this needs a verbier name.

File: lib/apivore/rspec_builder.rb [hunk 3]
```diff
@@ -1,114 +1,21 @@
...
-            expect(response).to have_http_status(response_code), "expected #{response_code} array, got #{response.status}: #{response.body}"
-
-            if fragment
-              expect(response.body).to conform_to_the_documented_model_for(swagger, fragment)
-            end
-
-          end
-        end
-      end
+  module RspecBuilder
```
Comment (gwagener) [inline lib/apivore/rspec_builder.rb]: :hocho:

File: lib/apivore/rspec_builder.rb [hunk 4]
```diff
@@ -1,114 +1,21 @@
 require 'apivore/rspec_matchers'
-require 'action_controller'
-require 'action_dispatch'
-require 'rspec/mocks'
+require 'apivore/document'
+require 'apivore/swagger_checker'
```
Comment (gwagener) [inline lib/apivore/rspec_builder.rb]: Not needed in here?

File: spec/request_spec.rb [hunk 5]
```diff
@@ -1,31 +1,50 @@
...
+      expect(subject).to document(
+        :patch, "/services/{id}.json", 204, {'id' => 1}
+      )
+    end
   end
 
-  apivore_setup '/services.json', 'get', '200' do
-    expect(Rails.application).to receive(:call).and_call_original
+  context "and" do
+    it { is_expected.to document_all_paths }
```
Comment (gwagener) [inline spec/request_spec.rb]: This generates "should have tested all documented routes" which is an older style. Current best practice is to have active sentences, for example "tests all documented routes". I don't think you can easily change the text `is_expected` generates so the best way to fix this is probably:
```ruby
context 'and' do
  it 'tests all documented routes' do
    expect(subject).to document_all_paths
  end
end
```

**Comment by gingermusketeer** [comment]
@hornc All the tests are passing but i can't see the status on this branch :( not sure if that is something you can fix.

I have changed deal service to use this api. You can see it in action here https://github.com/westfield/deal_service/pull/84

**Comment by jgalilee** [comment]
:+1: :100: :whale2: 

This looks fantastic to me. The refactor into an object will make debugging far easier.

File: lib/apivore/rspec_builder.rb (hunk 3 above)
Comment (jgalilee) [inline lib/apivore/rspec_builder.rb]: :sushi:

File: lib/apivore/swagger_checker.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    def load_swagger_doc!
+      @swagger = Apivore::Swagger.new(fetch_swagger!)
+    end
+
+    def fetch_swagger!
+      session = ActionDispatch::Integration::Session.new(Rails.application)
+      begin
+        session.get(swagger_path)
+      rescue
+        # TODO: make this fail inside rspec test execution rather than immediately raise an exception.
```
Comment (gwagener) [inline lib/apivore/swagger_checker.rb]: Still needed?

File: apivore.gemspec [hunk 7]
```diff
@@ -2,7 +2,7 @@ $:.push File.expand_path("../lib", __FILE__)
 
 Gem::Specification.new do |s|
   s.name        = 'apivore'
-  s.version     = '0.0.2'
+  s.version     = '1.0.0'
```
Comment (adamcohen) [inline apivore.gemspec]: I think this might be our first `1.0` gem! :+1:

File: lib/apivore/validator.rb [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    def pre_checks(swagger_checker)
+      check_request_path(swagger_checker)
+    end
+
+    def post_checks(swagger_checker)
+      check_status_code
+      check_response_is_valid(swagger_checker) unless has_errors?
+    end
+
+    def check_request_path(swagger_checker)
```
Comment (adamcohen) [inline lib/apivore/validator.rb]: It's great to see these methods split out like this, will make our life easier to track down errors

**Comment by adamcohen** [comment]
LGTM great work @gingermusketeer !

File: spec/data/example_specs.rb [hunk 9]
```diff
@@ -1,54 +1,64 @@
...
-  before do
-    apivore_setup '/services/{id}.json', 'get', '200' do
-      {'id' => 1}
-    end
+  include Apivore::RspecMatchers
+  include Apivore::RspecHelpers
 
-    apivore_setup '/services.json', 'post', '204' do
-      {"_data" => {'name' => 'hello world'}}
+  describe "undocumented path in test", :type => :request do
```
Comment (hornc) [inline spec/data/example_specs.rb]: This test is meant to confirm that a correctly documented path that is not yet implemented in the mock API will show up in the output.

does `describe "unimplemented path in API", :type => :request do` convey that?

There will need to be wording changes in spec/integration_spec.rb too

The current failure output for the test is:
`Swagger doc: /02_unimplemented_path.json does not have a documented path for /path_does_not_exist`

which is not what the test is trying to do -- it should be testing for the documented path `/not_implemented.json`, which will fail when that request is made to the mock API

File: spec/integration_spec.rb [hunk 10]
```diff
@@ -1,25 +1,19 @@
...
 
-  describe "the swagger documents an unimplemented path" do
-    it 'should show an alert that the specific path is unimplemented' do
-      stdout = `rspec spec/data/example_specs.rb --example 'unimplemented path in doc'`
+  describe "the test checks undocumented path" do
+    it 'should show an alert that the specific path is undocumented' do
+      stdout = `rspec spec/data/example_specs.rb --example 'undocumented path in test'`
       expect(stdout).to match(/1 failure/)
-      expect(stdout).to match(/expected 200 array, got 404/)
+      expect(stdout).to match(/does not have a documented path/)
```
Comment (hornc) [inline spec/integration_spec.rb]: The previous expectation of this test is the correct one  -- a successful 200 response containing an array is documented, but the mock API is returning a 404. 
`expect(stdout).to match(/expected 200 array, got 404/)`

The intent of this test was to be a clear indication that a route wasn't implemented.
Now that I think about it, this scenario should be covered by
*a path exists but the API responds with an unexpected http response code*

The existence  of the path or route doesn't really matter here. If that is the case, then this test can be removed, and the other scenario updated to:
*the API responds with an unexpected http response code* ? I'm thinking as I type here. Initially the intent was that an error in how a response was implemented was a different scenario to a path not existing at all..

File: spec/integration_spec.rb [hunk 11]
```diff
@@ -1,25 +1,19 @@
...
     end
   end
 
   describe "a path exists but the API response contains a property of a different type" do
     it 'should show which path and field has the problem for both index and view' do
```
Comment (hornc) [inline spec/integration_spec.rb]: `both index and view` seems to be no longer true here, it doesn't have an 'also fails' like some of the other scenarios

File: spec/integration_spec.rb (hunk 10 above)
Comment (gingermusketeer) [inline spec/integration_spec.rb]: I have fixed this. It does not say `200 array` anymore. I dont think it conveys what is intended. If the status codes do not match we should not check anything else

File: spec/data/example_specs.rb [hunk 12]
```diff
@@ -1,54 +1,64 @@
...
+    it "also fails" do
+      expect(subject).to validate(:get, "/services/{id}.json", 200, { "id" => 1})
+    end
   end
 
   describe "missing non-required", :type => :request do
-    validate("/07_missing_non-required_property.json")
+    subject { Apivore::SwaggerChecker.instance_for("/07_missing_non-required_property.json") }
+
+    it "fails" do
```
Comment (hornc) [inline spec/data/example_specs.rb]: should be "passes" here

File: spec/data/example_specs.rb [hunk 13]
```diff
@@ -1,54 +1,64 @@
...
+      expect(subject).to validate(:get, "/services.json", 200)
+    end
+
+    it "also fails" do
```
Comment (hornc) [inline spec/data/example_specs.rb]: and "also passes"

**Comment by jgalilee** [comment]
:100: :whale2: :+1:

**Comment by leondewey** [comment]
@gingermusketeer @hornc Im worried about this PR. I thought it was a syntax change not a relaxing/change in philosophy of the way we are testing. 

Can the three of us catch up before this get merged or has any more work done on it.

**Comment by hornc** [comment]
The method `validate_all_paths()` does not give accurate failure messages, and I think perhaps it is mis-named? `validate()` checks the swagger and makes requests to the API, I suggested the name `validate_all_paths()` because I thought it did a similar thing, because now I see that it only checks that all documented paths have been tested.

```
3) API testing scenarios unexpected http response fails
     Failure/Error: expect(subject).to validate_all_paths
       get /services.json is undocumented for response codes ["222", ["#", "paths", "/services.json", "get", "responses", "222", "schema"]]
     # ./spec/data/example_specs.rb:26:in `block (3 levels) in <top (required)>' 
```

In my test swagger there was a documented path for /services.json response code 222. I think the error was that there was no explicit test for it in prior tests.

`The documented path for get /services.json is untested for response codes ["222", ["#", "paths", "/services.json", "get", "responses", "222", "schema"]]` is more accurate? I could be wrong because this method confused me at first.

**Comment by gingermusketeer** [comment]
Yep. That is probably a better message. I think the main thing with this PR is that we want to do a good job with the api so people dont need to change it too often. Improving error messages can be done really easily post merge. If you can think of a better name that would be great. I will look at making these changes tomorrow.

**Comment by gingermusketeer** [comment]
@hornc I have tweaked the message when paths are untested.

I can't think of a better name for the validate all method...

**Comment by hornc** [comment]
re. `validate_all_paths` naming, after some more though and discussion with @leondewey, I think it can stay like that, and we can add some code to attempt to validate the not -specifically tested paths against the API, if it can (no substitutions etc required), and then spit out a report about what could not be tested, and possibly generate an rspec template from the swagger that can be copied back into the tests?

**Pull request merged**